### ISSUE TITLE
Compilation error during kASA make command

### ISSUE THREAD

**Issue by jfy133** [SilvioWeging/kASA#7]
Hello,

I was looking to trying out kASA, and was following the instructions to compile the tool myself (as the pre-built binary didn't work for me).

zLib compiled fine, but running the `make` of these series of instructions from your README:

```bash
cd <installPath>/build (or create/rename the folder)
cmake -DCMAKE_BUILD_TYPE=Release ..
make
```

Resulted in the following error

```
Scanning dependencies of target kASA                                                                                  
[ 95%] Building CXX object CMakeFiles/kASA.dir/source/main.cpp.o                                                      
In file included from /u/jfellows/bin/kASA/source/kASA.hpp:14:0,                                                      
                 from /u/jfellows/bin/kASA/source/main.cpp:10:                                                        
/u/jfellows/bin/kASA/source/utils/Utilities.hpp: In instantiation of ‘void kASA::Compare<contentVecType, elemType, int
Type>::CompareWithLib_partialSort(const string&, const string&, const string&, const string&, const string&, const uin
t8_t&, const int64_t&, const bool&, bool, const bool&, const uint8_t&, const float&, int32_t, const int32_t&) [with co
ntentVecType = stxxl::vector<packedBigPair, 4u, stxxl::lru_pager<4u>, 2101248u, stxxl::RC, long long unsigned int>; el
emType = packedBigPair; intType = long unsigned int; std::string = std::basic_string<char>; uint8_t = unsigned char; i
nt64_t = long int; int32_t = int]’:                                                                                   
/u/jfellows/bin/kASA/source/main.cpp:880:182:   required from here                                                    
/u/jfellows/bin/kASA/source/utils/Utilities.hpp:380:8: error: conversion from ‘const char [2]’ to non-scalar type ‘con
st string {aka const std::basic_string<char>}’ requested                                                              
  class FileReader {                                                                                                  
        ^                                                                                                             
In file included from /u/jfellows/bin/kASA/source/main.cpp:13:0:                                                      
/u/jfellows/bin/kASA/source/modes/Compare.hpp:2119:38: note: synthesized method ‘Utilities::FileReader<std::basic_ifst
ream<char> >::FileReader()’ first required here                                                                       
      Utilities::FileReader<ifstream> fileReaderObject;                                                               
                                      ^
In file included from /u/jfellows/bin/kASA/source/kASA.hpp:14:0,
                 from /u/jfellows/bin/kASA/source/main.cpp:10:
/u/jfellows/bin/kASA/source/utils/Utilities.hpp:380:8: error: conversion from ‘const char [2]’ to non-scalar type ‘const string {aka const std::basic_string<char>}’ requested
  class FileReader {
        ^
In file included from /u/jfellows/bin/kASA/source/main.cpp:13:0:
/u/jfellows/bin/kASA/source/modes/Compare.hpp:2120:39: note: synthesized method ‘Utilities::FileReader<igzstream>::FileReader()’ first required here 
      Utilities::FileReader<igzstream> fileReaderObject_gz; 
                                       ^
In file included from /u/jfellows/bin/kASA/source/main.cpp:11:0:
/u/jfellows/bin/kASA/source/modes/Read.hpp:273:10: error: conversion from ‘const char [1]’ to non-scalar type ‘std::string {aka std::basic_string<char>}’ requested
   struct strTransfer {
          ^
/u/jfellows/bin/kASA/source/modes/Read.hpp:273:10: error: conversion from ‘const char [1]’ to non-scalar type ‘std::string {aka std::basic_string<char>}’ requested
/u/jfellows/bin/kASA/source/modes/Read.hpp:273:10: error: conversion from ‘const char [1]’ to non-scalar type ‘std::string {aka std::basic_string<char>}’ requested
/u/jfellows/bin/kASA/source/modes/Read.hpp:273:10: error: conversion from ‘const char [1]’ to non-scalar type ‘std::string {aka std::basic_string<char>}’ requested
/u/jfellows/bin/kASA/source/modes/Read.hpp:273:10: error: conversion from ‘const char [1]’ to non-scalar type ‘std::string {aka std::basic_string<char>}’ requested
In file included from /u/jfellows/bin/kASA/source/main.cpp:13:0:
/u/jfellows/bin/kASA/source/modes/Compare.hpp:2237:95: note: synthesized method ‘constexpr kASA::Read<stxxl::vector<packedBigPair, 4u, stxxl::lru_pager<4u>, 2101248u, stxxl::RC, long long unsigned int>, packedBigPair, long unsigned int>::strTransfer::strTransfer()’ first required here 
      unique_ptr<typename Base::strTransfer> transferBetweenRuns(new typename Base::strTransfer);
                                                                                               ^
/u/jfellows/bin/kASA/source/modes/Compare.hpp:2447: confused by earlier errors, bailing out
CMakeFiles/kASA.dir/build.make:81: recipe for target 'CMakeFiles/kASA.dir/source/main.cpp.o' failed
make[2]: *** [CMakeFiles/kASA.dir/source/main.cpp.o] Error 1
CMakeFiles/Makefile2:160: recipe for target 'CMakeFiles/kASA.dir/all' failed
make[1]: *** [CMakeFiles/kASA.dir/all] Error 2
Makefile:170: recipe for target 'all' failed
make: *** [all] Error 2
````

Note the following loaded Modulefiles:

```
Currently Loaded Modulefiles:
  1) intel/19.1.1         3) mkl/2020.1           5) anaconda/3/2020.02   7) gcc/6
  2) impi/2019.7          4) jdk/11.0.5           6) clang/11.0           8) cmake/3.18
```

I used the GCC/clang/cmake versions _as far as could make out_ from the README (sorry if I missed something obvious, the README wasn't 100% clear to me, and I've got too used to bioconda nowadays it seems 😅)

Do you have any suggestions?

Cheers

**Comment by ghost** [comment]
Greetings!

Thank you for testing out my tool :)

I tried to fix the compiler errors in the current commit, please try again now. 
Furthermore: I would be glad if you could also attach the version of your compiler (in gcc this would mean the output of `gcc -v`) in the future. My best guess for a reason behind these errors is that your gcc version could not process strings being initialized as member variables with a non-scalar type (for other Nerds who read this: `class Example { string something = string(""); };` vs `class Example { string something = ""; };`).  

Would you also care to point me to the parts in the readme that are not precise enough? I'm quite aware that it is written from the perspective of the developer, not the user (but I aim to improve!) :D 

Cheers

**Comment by jfy133** [comment]
Hi Silvio,

Thanks for the reply!

Unfortunately it did not work. Find attached logs from `cmake` and `make` of kASA itself.
[kASA_make_fail.log](https://github.com/SilvioWeging/kASA/files/6341486/kASA_make_fail.log)
[kASA_cmake_success.txt](https://github.com/SilvioWeging/kASA/files/6341487/kASA_cmake_success.txt)



---

Oops, sorry. I naïvely thought the `gcc/6` in my module list would've be enough as the sys. admins here are normally pretty good with listing full version numbers ( but appears not in this case, my bad).

Output:

```bash
🕙 09:12:04 on cobra02 in kASA on  master ❯ gcc -v
Reading specs from /cobra/u/system/soft/SLE_12/packages/x86_64/gcc/6.4.0/bin/../lib/gcc/x86_64-pc-linux-gnu/6.4.0/specs
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/cobra/u/system/soft/SLE_12/packages/x86_64/gcc/6.4.0/bin/../libexec/gcc/x86_64-pc-linux-gnu/6.4.0/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: ../gcc-6.4.0/configure LDFLAGS='-fmessage-length=0 -grecord-gcc-switches -O2 -D_FORTIFY_SOURCE=2 -funwind-tables -fasynchronous-unwind-tables -U_FORTIFY_SOURCE -fPIC -Wno-error=format-security -Wl,-rpath=/mpcdf/soft/SLE_12/packages/x86_64/gcc/6.4.0/lib64 -L/mpcdf/soft/SLE_12/packages/x86_64/isl/0.16.1/lib -Wl,-rpath=/mpcdf/soft/SLE_12/packages/x86_64/isl/0.16.1/lib' CFLAGS='-fmessage-length=0 -grecord-gcc-switches -O2 -D_FORTIFY_SOURCE=2 -funwind-tables -fasynchronous-unwind-tables -U_FORTIFY_SOURCE -fPIC -Wno-error=format-security' CXXFLAGS='-fmessage-length=0 -grecord-gcc-switches -O2 -D_FORTIFY_SOURCE=2 -funwind-tables -fasynchronous-unwind-tables -U_FORTIFY_SOURCE -fPIC -Wno-error=format-security' XCFLAGS='-fmessage-length=0 -grecord-gcc-switches -O2 -D_FORTIFY_SOURCE=2 -funwind-tables -fasynchronous-unwind-tables -U_FORTIFY_SOURCE -fPIC -Wno-error=format-security' TCFLAGS='-fmessage-length=0 -grecord-gcc-switches -O2 -D_FORTIFY_SOURCE=2 -funwind-tables -fasynchronous-unwind-tables -U_FORTIFY_SOURCE -fPIC -Wno-error=format-security' GCJFLAGS='-fmessage-length=0 -grecord-gcc-switches -O2 -D_FORTIFY_SOURCE=2 -funwind-tables -fasynchronous-unwind-tables -U_FORTIFY_SOURCE -fPIC -Wno-error=format-security ' --with-stage1-ldflags='-fmessage-length=0 -grecord-gcc-switches -O2 -D_FORTIFY_SOURCE=2 -funwind-tables -fasynchronous-unwind-tables -U_FORTIFY_SOURCE -fPIC -Wno-error=format-security -Wl,-rpath=/mpcdf/soft/SLE_12/packages/x86_64/gcc/6.4.0/lib64 -L/mpcdf/soft/SLE_12/packages/x86_64/isl/0.16.1/lib -Wl,-rpath=/mpcdf/soft/SLE_12/packages/x86_64/isl/0.16.1/lib' --prefix=/mpcdf/soft/SLE_12/packages/x86_64/gcc/6.4.0 --disable-multilib --with-isl=/mpcdf/soft/SLE_12/packages/x86_64/isl/0.16.1 --enable-linker-build-id --enable-languages=c,c++,fortran
Thread model: posix
gcc version 6.4.0 (GCC) 
🕙 09:12:57 on cobra02 in kASA on  master ❯ clang --version
clang version 11.0.0
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /mpcdf/soft/SLE_12/packages/x86_64/clang/11.0.0/bin
🕙 09:13:20 on cobra02 in kASA on  master ❯ cmake --version
cmake version 3.18.4

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```
---

Suggestion: maybe you could make a bug report issue template with a list of requested output  with recommended commands? We've used [that](https://github.com/nf-core/eager/blob/master/.github/ISSUE_TEMPLATE/bug_report.md) in one of our own projects and it works quite nicely to make sure unfamiliar users (or slightly lazy, like me - sorry) more likely put the required info for debugging.

---

The bit of the README that wasn't 100% clear was this bit:

> If you need to compile the code, you'll definitely need a C++ compiler that supports C++11 or if possible C++17 (for filesystem and execution). I successfully tested it with Visual Studio 2017/2019 (using C++17 is necessary for this since experimental\filesystem is deprecated), GCC version 6.1, LLVM/Clang 9.0 and Apple Clang 9.0. Here is a table showing if your version is sufficient. On Linux and macOS, cmake is needed as well.

As I'm not that familiar with C++, I wasn't really sure what the  `A C++ compiler that supports C++11 <...>` would really mean, and also a version of `cmake` wasn't specified.I didn't find the table you link to that helpful either unfortunately (didn't really know what I was looking at).

I might be being a picky, but my suggestion would also suggest maybe change that paragraph into a list of bullet-points (or even a table to account for different OSs), as it makes it easier to use a 'checklist' that a user got everything.

Cheers,

**Comment by ghost** [comment]
Hey James,

Thank you very much for these files. I see that cmake is not using the correct gcc/g++ version (4.8) although you have version 6.4 installed. Please modify the call to cmake to `cmake -DCMAKE_BUILD_TYPE=Release -D CMAKE_C_COMPILER=gcc-6 -D CMAKE_CXX_COMPILER=g++-6 ..`. I tried it in my WSL and all gcc/g++ versions starting from 5 are working but 4 does not.

Also, thank you for the suggestion on the bug report templates. This is a tremendous help!

Regarding the README: Your critique is appreciated, I changed the README and wiki accordingly.

Cheers
Silvio

**Comment by jfy133** [comment]
Wonderful! 

I have now got it to compile (in this case actually with gcc v8.4, as you noted that would also work, and I assumed latest  == better)! Thank you very much!

I also made a PR to fix a formatting issue in #8  

Thank you for being so responsive!
